recipes/docker-image-guide.md

```diff
@@ -0,0 +1,201 @@
+# Docker image guide
+
+The following handbook offers best practices for creating small and secure NodeJs
+Docker images suitable for production use.
+You will find it helpfull no matter what type of NodeJs application you aim to build.
+
+Examples with each step explained will be used to guide you through best practices.
+
+## Simple NodeJs application
+
+### The application
+
+Let's start with the simple nodejs application. Here is an overview of the files
+included in the project:
+
+```
+├── index.js
+├── package.json
+├── package-lock.json
+├── Dockerfile
+├── .dockerignore
+```
+
+```js
+// index.js
+const express = require("express");
+const os = require("os");
+
+const app = express();
+
+app.use("/", (req, res) => {
+  res.send(`Hello world from ${os.hostname()}.`);
+});
+
+app.listen(3000, () => {
+  console.log("App is listening on port 3000");
+});
+```
+
+### The Dockerfile
+
+```dockerfile
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#1-use-explicit-and-deterministic-docker-base-image-tags
+# https://snyk.io/blog/choosing-the-best-node-js-docker-image
+FROM node:20.8.1-bookworm-slim@sha256:01e0a13c8c2daecd73908e922207e9695a82d9d41019a96628751fa9f7c0d853 AS configure
+# https://docs.docker.com/build/cache/#combine-commands-together-wherever-possible
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#5-properly-handle-events-to-safely-terminate-a-nodejs-docker-web-application
+# https://github.com/Yelp/dumb-init
+RUN apt update && apt install dumb-init
+ENTRYPOINT ["dumb-init", "--"]
+
+FROM configure AS install
+# https://www.pathname.com/fhs/pub/fhs-2.3.html#USRSRCSOURCECODE
+WORKDIR /usr/src/app
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#3-optimize-nodejs-tooling-for-production
+ENV NODE_ENV production
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#4-dont-run-containers-as-root
+COPY --chown=node:node package*.json .
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#2-install-only-production-dependencies-in-the-nodejs-docker-image
+RUN npm ci --only=production
+COPY --chown=node:node . .
+
+FROM install AS run
+# https://cheatsheetseries.owasp.org/cheatsheets/NodeJS_Docker_Cheat_Sheet.html#4-dont-run-containers-as-root
+USER node
+CMD [ "node", "index.js" ]
```

Comment:
Is this not the recommended way of running the image so that the process is not ran as PID 1?

CMD ["dumb-init", "node", "server.js"]

Reply:
We set `ENTRYPOINT ["dumb-init", "--"]` which takes whatever we pass as cmd as an argument. The reason why we didn't do that inside CMD is because CMD can be overridden when starting the docker container.